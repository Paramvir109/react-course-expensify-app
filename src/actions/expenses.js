//component calls action gen
//action gen return obj
//component dispatches obj
//redux store changes

//To use firebase -> action gen returns a function
//component dispatches fn
//fn runs(can dispatch other actions and do whatever it wants)

import uuid from 'uuid';
import database from '../firebase/firebase'

// ADD_EXPENSE
export const addExpense =  (expense) => ({
    type : 'ADD_EXPENSE',
    expense
})

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {//We can't return functions as such (SO we used middleware, see store config)
         const   {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0

        } = expenseData;
        const expense = {description, note, amount, createdAt};
        return database.ref('expenses').push(expense).then((ref) =>{//thenable will recieve ref
            dispatch(addExpense({
                id : ref.key,
                ...expense
            }))
        })

    }
}


// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type : 'EDIT_EXPENSE',
    id,
    updates
})
// REMOVE_EXPENSE
export  const removeExpense = ({id} = {}) => ({
    type : 'REMOVE_EXPENSE',
    id
})      