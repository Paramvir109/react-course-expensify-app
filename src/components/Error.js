import { Link } from 'react-router-dom'
import React from 'react'

 //Link tag causes it to rerender(Not a full page refresh)
 const NotFoundPage = () => (
    <div>
      404! - <Link to="/">Go home</Link>
    </div>
);
  export default NotFoundPage;