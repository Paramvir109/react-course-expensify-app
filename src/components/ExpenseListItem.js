import React from 'react'
import { connect } from 'react-redux';
import { NavLink } from 'react-router-dom';

//dispatch method available from connect
export const ExpenseListItem = ({description, createdAt, amount, id, dispatch}) => (
    <div>
    <NavLink activeClassName="is-active" to={"/edit/" + id}>
         <h3>{description}</h3>
    </NavLink>
        <p>{amount} - {createdAt}</p>
       
    </div>
);



export default connect()(ExpenseListItem);