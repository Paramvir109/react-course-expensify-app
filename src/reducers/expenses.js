const expenseReducerDefaultState = []


const expenseReducer = (state = expenseReducerDefaultState, action) => {
    switch(action.type) {
        case 'ADD_EXPENSE' :
            return [...state,action.expense]//Not mutating state
        case 'REMOVE_EXPENSE' :
            return state.filter((expense) => expense.id !== action.id)
        case 'EDIT_EXPENSE' :
            return state.map((expense) => {
                if(expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates//spread op is also used to override existing key-value pairs
                    }
                } else {
                    return expense
                }
            })
        case 'SET_EXPENSES' : 
            return action.expenses
            
        default :
            return state
    }
};

export default expenseReducer;