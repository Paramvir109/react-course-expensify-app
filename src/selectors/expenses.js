import moment from 'moment';

export default (expenses, {startDate, endDate, text, sortBy}) => {
    //for filtering all expenses b/w a given duration and matched text in desccription field
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt)
        const startTimeMatch = startDate ? createdAtMoment.isSameOrAfter(startDate,'day') :true;
        const endTimeMatch = endDate ? createdAtMoment.isSameOrBefore(endDate, 'day') :true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());//Both converted to lowercase 
        return startTimeMatch && endTimeMatch && textMatch;
    }).sort((a,b) => {
        if(sortBy === 'date') {
           return a.createdAt < b.createdAt? 1 : -1;//Most recent expense first
        }
        if(sortBy === 'amount') {
            return a.amount < b.amount? 1 : -1;//Most expensive first
         }
    })
    
} 