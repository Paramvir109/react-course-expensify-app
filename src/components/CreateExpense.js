import React from 'react'
import { connect } from 'react-redux';
import {addExpense} from './../actions/expenses'
import ExpenseForm from './ExpenseForm';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.addExpense(expense);
    this.props.history.push('/');
  }
  render() {
    return (
      <div>
        <h1>This is from my add expense component</h1>
        <ExpenseForm onSubmit={this.onSubmit}/>
      </div>
    );
  }
}

// export const AddExpensePage = (props) => (
//     <div>
//       <h1>This is from my add expense component</h1>
//       <ExpenseForm 
//         onSubmit={(expense) => {
//           // props.dispatch(addExpense(expense)); will change it to make test-able as well
//           props.onSubmit(expense);
//           props.history.push('/');//Will automatically redirect to dashboard page
//         }}
        
//       />
//     </div>
// );

//optional -> used to dispatch actions
const mapDispatchToProps = (dispatch) => {
  return {
    addExpense : (expense) => dispatch(addExpense(expense))
  }
}

export default connect(undefined, mapDispatchToProps)(AddExpensePage);