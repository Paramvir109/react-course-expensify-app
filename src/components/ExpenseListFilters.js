import React from 'react'
import { connect } from 'react-redux';
import { setTextFilter,sortByAmount,sortByDate, setEndDate, setStartDate} from './../actions/filters';
import 'react-dates/initialize';
import uuid from 'uuid'

import {DateRangePicker} from 'react-dates';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused : null
    }
    onDatesChange = ({startDate, endDate}) => {
        this.props.setEndDate(endDate);
        this.props.setStartDate(startDate);

    }
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }))
    }
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
    }
    onSortChange = (e) => {
        this.props.filters.sortBy === 'date'? this.props.sortByAmount() : this.props.sortByDate()
    }

    render() {
       return (
            <div>
                <input 
                    type="text" 
                    value={this.props.filters.text} 
                    onChange={this.onTextChange}>
                </input>
                <select 
                    value={this.props.filters.sortBy} 
                    onChange={this.onSortChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker 
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    startDateId={uuid()}
                    endDateId={uuid()}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    showClearDates={true}
                />
            </div>
        );
    }
}

//props above contains two things 1.) filters 2.) dispatch method(bydefault)
const mapStateToProps = (state) => {
    return {
        filters : state.filters
    }
}
const mapDispatchToProps = (dispatch) => {
    return {
      setTextFilter : (value) => dispatch(setTextFilter(value)),
      setStartDate : (startDate) => dispatch(setStartDate(startDate)),
      setEndDate : (endDate) => dispatch(setTextFilter(endDate)),
      sortByAmount : () => dispatch(sortByAmount()),
      sortByDate : () => dispatch(sortByDate()),
    }
  }

export default connect(mapStateToProps,mapDispatchToProps)(ExpenseListFilters);