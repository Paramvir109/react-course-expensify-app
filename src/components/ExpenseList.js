import React from 'react'
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenseList from './../selectors/expenses';
export const ExpenseList = (props) => (//Route sends some built in props
    <div>
        <h1>Expense List</h1>
        {props.expenses.length === 0 ? (<p>No expenses</p>):(props.expenses.length)}
        {props.expenses.map((expense) => (<ExpenseListItem key={expense.id} {...expense}/>))}
    </div>
);
//return value of connect is the function which is to be called to create a hoc

// const connectedExpenseList = connect((state) => {
//     //here this function tells bout which properties of store can be accesssed by the component(connectedExpenseList)
//     return {
//         expenses : state.expenses
//     }

// })(ExpenseList);

// export default connectedExpenseList;

//OR
const mapStateToProps = (state) => {
    return {
        expenses : selectExpenseList(state.expenses, state.filters)
    }

}
export default connect(mapStateToProps)(ExpenseList);

