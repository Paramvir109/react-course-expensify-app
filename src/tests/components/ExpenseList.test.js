import React from 'react';
import { ExpenseList } from '../../components/ExpenseList';
import { shallow  } from 'enzyme';
import expenses from '../fixtures/expenses';

test('Should render expense list with expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
})
test('Should render expense list with empty messagee', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
})