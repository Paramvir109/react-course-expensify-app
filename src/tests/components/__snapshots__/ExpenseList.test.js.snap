// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render expense list with empty messagee 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <p>
    No expenses
  </p>
</div>
`;

exports[`Should render expense list with expenses 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  3
  <Connect(ExpenseListItem)
    amount={100}
    createdAt={0}
    description="Bill"
    id="1"
    key="1"
    note=""
  />
  <Connect(ExpenseListItem)
    amount={122000}
    createdAt={345600000}
    description="Rent"
    id="2"
    key="2"
    note=""
  />
  <Connect(ExpenseListItem)
    amount={1020000}
    createdAt={-345600000}
    description="Credit card"
    id="1"
    key="1"
    note=""
  />
</div>
`;
